pkg/passkey/passkey.go

```diff
@@ -0,0 +1,123 @@
+// Package passkey implements the support of Webauthn credentials for authentication.
+// It is based on the W3C's "Web Authentication: An API for accessing Public Key Credentials"
+// https://www.w3.org/TR/webauthn-2/
+package passkey
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+)
+
+type CredentialFlags struct {
+	UserPresent     bool
+	UserVerified    bool
+	AttestationData bool
+	ExtensionData   bool
+}
+
+// Structure to store information and key of public yey credential
+type Credential struct {
+	Id        string
+	Rp        string
+	UserName  string
+	Algorithm string
+	KeyValue  *ecdsa.PrivateKey
+	Counter   uint32
+	Flags     CredentialFlags
+}
+
+// Client data for signature: https://www.w3.org/TR/webauthn-1/#sec-client-data
+type ClientData struct {
+	Challenge string `json:"challenge"`
+	Origin    string `json:"origin"`
+	CredType  string `json:"type"`
+}
+
+// Response from a GetAssertion: https://www.w3.org/TR/webauthn-1/#authenticatorGetAssertion-return-values
+type Response struct {
+	AuthenticatorData string
+	ClientDataJSON    string
+	Signature         string
+	UserHandle        string
+}
+
+func authDataFlags(options CredentialFlags) uint8 {
+	flags := uint8(0)
+
+	if options.ExtensionData {
+		flags |= 0b10000000
+	}
+
+	if options.AttestationData {
+		flags |= 0b01000000
+	}
+
+	if options.UserVerified {
+		flags |= 0b00000100
+	}
+
+	if options.UserPresent {
+		flags |= 0b00000001
+	}
+
+	return flags
+}
+
+func MakeCredential(rp string, user string, flags CredentialFlags) (*Credential, error) {
+	rawId := make([]byte, 32)
+	_, err := rand.Read(rawId)
+	if err != nil {
+		return nil, fmt.Errorf("error while generating random ID: %w", err)
+	}
+	privateKey, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+
+	return &Credential{
+		Id:        base64.RawURLEncoding.EncodeToString(rawId),
+		Rp:        rp,
+		UserName:  user,
+		Algorithm: "ECDSA",
+		KeyValue:  privateKey,
+		Counter:   0,
+		Flags:     flags,
+	}, nil
+}
+
+func GetAssertion(challenge string, origin string, cred *Credential) (*Response, error) {
```

Comment:
I wonder if this shouldn't rather be a method with a *Credential receiver?

Reply:
It was to keep the same name as in the Webauthn specification: https://www.w3.org/TR/webauthn-2/#sctn-op-get-assertion

Reply:
I meant 
```suggestion
func (cred *Credential) GetAssertion(challenge string, origin string) (*Response, error) {
```

as in https://go.dev/tour/methods/4